pyteal/compiler/sourcemap.py

```diff
@@ -0,0 +1,1204 @@
+        pyteal.compiler.compiler.Copmilation._compile_impl()
+
+    OUTPUT: (self._cached_tmis, _cached_r3sourcemap, self._cached_pc_sourcemap)
+
+    PASS I. Deduce the Best Frame Candidate (BFC) from each individual `NatalStackFrame`
+
+    NOTE: this logic actually occurs at creation of the PyTeal Expr which is responsible
+    for each teal component. The `build()` method simply retrieves the result of this
+    pre-computation through NatalStackFrame.best()
+
+    for each component in components:
+        # Deduce the "best" frame for the component's PyTeal source (@ Expr creation)
+
+        # [1-7] Inside NatalStackFrame.__init__()
+        1. call inspect.stack() to generate a full stack trace
+        2. filter out "py crud" frames whose filename starts with "<" and don't have a code_context
+        3. start searching at frame index i = 2 (right before Expr's NatalStackFrame was constructed)
+        4. fast forward through PyTeal frames until the first non PyTeal frame is found
+        5. in the case this was in import statement, back up until a known compiler generated line is discovered
+        6. keep the last frame in the list
+        7. convert this into a list[StackFrame] of size 1
+
+        # [8] Inside _PyTealSourceMapper.build() @ source-map creation:
+        8. self._best_frames[i] = the singleton StackFrame in 7 converted to PyTealFrame  # i == component's index of the component
+
+    PASS II. Attempt to fill any "gaps" by inferring from adjacent BFC's
+    This logic is contained in _PyTealSourceMapper.infer():
+
+    NOTE: the mutations of self._best_framed described below are "sticky" in the sense
+    that when this `pyteal_frame` is modified and referenced by `prev_pyteal_frame`
+    in the next interation, it is the new _modified_ version that is used.
+
+    for each pyteal_frame in self._best_frames:
+        status = pyteal_frame.status()
+        if status NOT in PyTealFrameStatus.{MISSING, MISSING_AST, MISSING_CODE, PYTEAL_GENERATED}
+            NO-OP # we can't do any better
+
+        else: # call local method _PyTealSourceMapper.infer.infer_source()
+            prev_pyteal_frame, next_pyteal_frame <--- get these from self._best_frames (or None if doesn't exist)
+            if both exist:
+                if both are equal (have the equal `frame_info, node, rel_paths`):
+                    replace this pyteal_frame by next_pyteal_frame
+
+                if pyteal_frame is flagged as compiler generated:
+                    if its reason is one of PT_GENERATED.{TYPE_ENUM_ONCOMPLETE, TYPE_ENUM_TXN, BRANCH_LABEL, BRANCH}:
+                        # these occur when one of the # T2PT* comments is encountered
+                        replace this pyteal_frame by next_pyteal_frame
+                    if its reason is PT_GENERATED.FLAGGED_BY_DEV:
+                        # this occurs when NatalStackFrame._compiler_gen is set True
+                        replace this pyteal_frame by prev_pyteal_frame
+
```

Comment:
This is just a heuristic and is unlikely to always be correct, unfortunately. The best way to improve the mapping is to avoid these "misses" as much as possible, and that is what #676 aims to achieve.

Reply:
no worries, I am totally fine with it, and by a single PR I don't expect we can perfectly correctly fill the gaps.

Thanks for showing me #676 and I will focus there later.

Reply:
However, there is a need to look again at this code at the end of #676 and assess which of these hacks is still necessary. I've added a line item in #681 

<img width="640" alt="image" src="https://user-images.githubusercontent.com/291133/221975025-45790cdd-5d59-4946-a992-831b5b776f00.png">